### PR TITLE
Add NFTokenPages to account_objects RPC

### LINKED ISSUE #4347: account_objects method does not return NFTokenPages (Version: 1.9.4)
## Issue Description
When trying to get an xrpl accounts owned objects, the method `account_objects` does not return NFTokenPage(s). This is a problem because one could not reliably show the user which objects currently count towards his owner reserve.

The documentation also states:
`The account_objects command returns the raw [ledger format](https://xrpl.org/ledger-object-types.html) for all objects owned by an account.`

## Steps to Reproduce
1. call the `account_objects` method of an account which owns some NFTs
2. response does not include any NFTokenPage

## Expected Result
Also NFTokenPages are returned through the `account_objects` method.

## Actual Result
No NFTokenPage is returned

This is actually quite a problem. On accounts with NFTs, the OwnerCount of the AccountRoot objects shows a bigger number than objects are returned by the `account_objects` method. This confuses developers as well as users.

### PR DESCRIPTION
**Pull request by RichardAH**
## High Level Overview of Change
Fix #4347

NFTokenPages consume owner count and therefore should be returned by the `account_objects` RPC, lest users be unaware where their owner count has been allocated.

Credit to @scottschurr for unit test / bug fixes and polishing.

It is technically possible to iterate NFTokenPages using existing RPC calls, but it is hacky to do this, requires additional calls and increases complexity (without gain) for wallet / app developers. Therefore I strongly advise this PR is merged. 

### Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1203482814597350

### CODE REVIEW

**Comment by RichardAH** [comment]
```
Longest suite times:
   43.0s ripple.tx.Offer
   35.1s ripple.app.ValidatorSite
   21.7s ripple.rpc.NodeToShardRPC
   21.2s ripple.tx.NFToken
   18.4s ripple.tx.NFTokenBurn
   14.0s ripple.app.Flow
   14.0s ripple.app.ShardArchiveHandler
   13.6s ripple.app.TheoreticalQuality
    8.8s ripple.tx.NFTokenDir
    8.2s ripple.app.LedgerReplayer
341.8s, 204 suites, 1528 cases, 532270 tests total, 0 failures
```

**Comment by cjcobb23** [comment]
Why exactly is it hacky to do this client side? We have the `account_nfts` RPC for this: https://xrpl.org/account_nfts.html. Maybe I am misunderstanding something though.

**Comment by RichardAH** [comment]
> Why exactly is it hacky to do this client side? We have the `account_nfts` RPC for this: https://xrpl.org/account_nfts.html. Maybe I am misunderstanding something though.

`account_nfts` does not tell you anything about how many NFT _pages_ you have and thus nothing about how much ownercount has been reserved by those pages. Everything that contributes to your ownercount should be displayed in `account_objects`.

**Comment by cjcobb23** [comment]
> > Why exactly is it hacky to do this client side? We have the `account_nfts` RPC for this: https://xrpl.org/account_nfts.html. Maybe I am misunderstanding something though.
> 
> `account_nfts` does not tell you anything about how many NFT _pages_ you have and thus nothing about how much ownercount has been reserved by those pages. Everything that contributes to your ownercount should be displayed in `account_objects`.

Does `account_objects` tell you about pages? It looks like `account_objects` just shows the raw ledger objects themselves, and doesn't say anything about the pages those objects are in.

**Review by ximinez — CHANGES_REQUESTED** [review]
The only real dealbreaker here is dereferencing the unseated optional.

File: src/test/rpc/AccountObjects_test.cpp [hunk 1]
```diff
@@ -333,6 +335,197 @@ class AccountObjects_test : public beast::unit_test::suite
...
+            Json::Value params;
+            params[jss::account] = bob.human();
+            params[jss::limit] = 1;
+
+            Json::Value resp =
+                env.rpc("json", "account_objects", to_string(params));
+            Json::Value& aobjs = resp[jss::result][jss::account_objects];
+            BEAST_EXPECT(aobjs.size() == 1);
+            auto& aobj = aobjs[0U];
+            BEAST_EXPECT(!resp[jss::result].isMember(jss::limit));
```
[the remark this answers is not in the document]
Comment (RichardAH) [reply src/test/rpc/AccountObjects_test.cpp:415]: @scottschurr

**Comment by RichardAH** [comment]
@ximinez I added a fix for the unguarded std::optional. As to the unit tests, those were written by @scottschurr so I will let him discuss. Unfortunately I don't have time at the moment.

**Review by scottschurr — COMMENTED** [review]
Sorry to be slow responding to the earlier comments.

I've incorporated the changes suggested by @ximinez in the top commit of this branch: https://github.com/scottschurr/rippled/commits/RichardAH-nftpage_accobjs. @RichardAH, if you could cherry pick that commit onto this branch then I believe all of @ximinez's comments will be addressed.  Thanks.

**Comment by intelliot** [comment]
@RichardAH this is ready for your perusal, at your convenience

**Comment by RichardAH** [comment]
Sorry for the late reply. Cherry-picked @scottschurr 's commit (thanks!) 
Looks like this still merge-conflict-free so whenever you want to merge it is fine by me

**Comment by intelliot** [comment]
This requires at least 2 code reviews. @HowardHinnant @scottschurr could you review, or suggest other reviewer(s)?

**Review by scottschurr — APPROVED** [review]
:+1: LGTM.  It might be good to have @ximinez give it a quick looking over so he can make sure all the comments he made have been addressed.  But as far as I'm concerned it's good to go.

File: src/ripple/rpc/impl/RPCHelpers.cpp [hunk 2]
```diff
@@ -171,25 +252,27 @@ getAccountObjects(
             found = true;
         }
 
+        // it's possible that the returned NFTPages exactly filled the
+        // response.  Check for that condition.
+        if (i == mlimit && mlimit < limit)
```
Comment (shawnxie999) [inline src/ripple/rpc/impl/RPCHelpers.cpp:257]: curious why `i == mlimit` was preferred here over than writing `0 == mlimit`

**Review by shawnxie999 — APPROVED** [review]
looks good

File: src/test/rpc/AccountObjects_test.cpp [hunk 3]
```diff
@@ -376,12 +579,29 @@ class AccountObjects_test : public beast::unit_test::suite
...
+        {
+            // Find the NFToken page and make sure it's the right one.
+            Json::Value const resp = acct_objs(gw, jss::nft_page);
+            BEAST_EXPECT(acct_objs_is_size(resp, 1));
+
+            auto const& nftPage = resp[jss::result][jss::account_objects][0u];
+            BEAST_EXPECT(nftPage[sfNFTokens.jsonName].size() == 1);
+            BEAST_EXPECT(
+                nftPage[sfNFTokens.jsonName][0u][sfNFToken.jsonName]
+                       [sfNFTokenID.jsonName] == to_string(nftID));
```
Comment (shawnxie999) [inline src/test/rpc/AccountObjects_test.cpp:602]: @scottschurr just curious isnt `string` comparison more expensive than comparing in `uint256`?

File: src/ripple/rpc/impl/RPCHelpers.cpp (hunk 2 above)
Comment (scottschurr) [reply src/ripple/rpc/impl/RPCHelpers.cpp:257]: It's been a while.  As I recall this change was motivated by one of the new unit tests (that failed before this change).  FWIW, `i` is incrementing inside the outer `for` loop, so I suspect that a comparison to `0` would not give the correct results.

File: src/test/rpc/AccountObjects_test.cpp [hunk 4]
```diff
@@ -510,7 +730,7 @@ class AccountObjects_test : public beast::unit_test::suite
             auto const& ticket = resp[jss::result][jss::account_objects][0u];
```
[the remark this answers is not in the document]
Comment (shawnxie999) [reply src/test/rpc/AccountObjects_test.cpp:730]: oh sorry, i was just pointing out the CI failure, which seemed to have failed on both ubuntu and mac

File: src/ripple/rpc/impl/RPCHelpers.cpp (hunk 2 above)
Comment (shawnxie999) [reply src/ripple/rpc/impl/RPCHelpers.cpp:257]: ah sorry i misread the code.

File: src/test/rpc/AccountObjects_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottschurr) [reply src/test/rpc/AccountObjects_test.cpp:730]: Thanks for the clarification @shawnxie999.  I am seeing two CI unit tests failures currently:

-  Ubuntu: `#65 failed: AccountLinesRPC_test.cpp(730)`
- Mac: `#65 failed: AccountLinesRPC_test.cpp(732)`

AccountLinesRPC_test.cpp has a file name that looks a lot like this file, but it's not the same file.  When I look in the source for that file the reported failing lines don't align with a `BEAST_EXPECT`.  So I'm inclined to suspect those are just two more flaky failures.  I wish that were not the case...

[the remark this answers is not in the document]
Comment (shawnxie999) [reply src/test/rpc/AccountObjects_test.cpp:730]: Thanks for the clarification. It's my first time seeing a flaky ubuntu test, if it is the case. All good.

**Comment by intelliot** [comment]
> It might be good to have @ximinez give it a quick looking over so he can make sure all the comments he made have been addressed.

**Review by HowardHinnant — APPROVED** [review]
Looks good.

**Review by ximinez — CHANGES_REQUESTED** [review]
Everything looks good from my earlier comments. One unit test is failing right now, due to interactions with my earlier PR #4361. Fortunately, I noticed the incompatibility at the time, and coded up a fix. Cherry pick this commit, and you should be good to go: https://github.com/ximinez/rippled/commit/b633d474e909668f93b7b79172be0e9a41753565

**Review by ximinez — APPROVED** [review]
Looks good to go.

**Comment by intelliot** [comment]
Related Clio PR: https://github.com/XRPLF/clio/pull/780

**Pull request merged**